subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+
+	/* Start looking from root directory of file system */
+	rc = ext2_inode_get(fs, EXT2_ROOT_INODE, &cur_dir);
+	if (rc < 0) {
+		ret = rc;
+		goto out;
+	}
+
+	/* There may be slash at the beginning of path */
+	const char *path = args->path;
+
+	path = skip_slash(path);
+
+	/* If path is empty then return root directory */
+	if (path[0] == '\0') {
+		args->inode = cur_dir;
+		cur_dir = NULL;
+		ret = 0;
+		goto out;
+	}
+
+	for (;;) {
+		/* Get path component */
+		char *end = strchrnul(path, '/');
+		size_t len = end - path;
+
+		if (len > EXT2_MAX_FILE_NAME) {
+			ret = -ENAMETOOLONG;
+			goto out;
+		}
+
+		strncpy(name_buf, path, len);
+		name_buf[len] = '\0';
+
+		/* Search in current directory */
+		uint32_t dir_off = 0;
+		/* using 64 bit value to don't lose any information on error */
+		int64_t ino = find_dir_entry(cur_dir, name_buf, len, &dir_off);
+
+		const char *next_path = skip_slash(end);
+		bool last_entry = next_path[0] == '\0';
+
+		if (!last_entry) {
+			/*  prepare the next loop iteration */
+
+			if (ino < 0) {
+				/* next entry not found */
+				ret = -ENOENT;
+				goto out;
+			}
+
+			rc = ext2_inode_get(fs, ino, &next);
+			if (rc < 0) {
+				/* error while fetching next entry */
+				ret = rc;
+				goto out;
+			}
+
+			if (!(next->i_mode & EXT2_S_IFDIR)) {
+				/* path component should be directory */
+				ret = -ENOTDIR;
+				goto out;
+			}
+
+			/* Go to the next path component */
+			path = next_path;
+
+			/* Move to next directory */
+			ext2_inode_drop(cur_dir);
+			cur_dir = next;
+
+			next = NULL;
+			continue;
+		}
+
+		/* Last entry */
+
+		if (ino < 0 && !(args->flags & LOOKUP_ARG_CREATE)) {
+			/* entry not found but we need it */
+			ret = -ENOENT;
+			goto out;
+		}
+
+		if (ino > 0) {
+			rc = ext2_inode_get(fs, ino, &next);
+			if (rc < 0) {
+				ret = rc;
+				goto out;
+			}
+		}
+
+		/* Store parent directory and offset in parent directory */
+		if (args->flags & (LOOKUP_ARG_CREATE | LOOKUP_ARG_STAT | LOOKUP_ARG_UNLINK)) {
+			/* In create it will be valid only if we have found existing file */
+			args->offset = dir_off;
+			args->parent = cur_dir;
+			cur_dir = NULL;
+		}
+
+		/* Store name info */
+		if (args->flags & LOOKUP_ARG_CREATE) {
+			args->name_pos = path - args->path;
+			args->name_len = len;
+		}
+
+		/* Store found inode */
+		if (ino > 0) {
+			args->inode = next;
+			next = NULL;
+		}
+
+		ret = 0;
+		goto out;
+	}
+
+out:
+	/* Always free that inodes.
+	 * If some of them is returned from function then proper pointer is set to NULL.
+	 */
+	ext2_inode_drop(cur_dir);
+	ext2_inode_drop(next);
+	return ret;
+}
+
+static const char *skip_slash(const char *str)
+{
+	while (*str != '\0') {
+		if (*str != '/') {
+			return str;
+		}
+		str++;
+	}
+	return str;
+}
+
+/* Return position of given char or end of string. */
+static char *strchrnul(const char *str, const char c)
+{
+	char *p = strchr(str, c);
+
+	if (!p) {
+		p = strchr(str, '\0');
```

Comment:
`strlen()`? The `strchr` checks for both terminating 0 and character you make it search.
If both `strchr` fail here you basically run twice full length of str doing two comparisons each run.
Would be more efficient to just write `strchrnul` from scratch.

Reply:
I've implemented strchrnul to be used here.